src/test.ts

```diff
@@ -0,0 +1,163 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {SpawnOptions} from 'child_process';
+import * as packlist from 'npm-packlist';
+import * as path from 'path';
+import * as tar from 'tar';
+
+import {globP, mkdirP, rimrafP, spawnP, tmpDirP, writeFileP} from './utils';
+
+const INDEX_TS = 'index.ts';
+const INDEX_JS = 'index.js';
+
+export interface BaseCodeSample {
+  description: string;
+  dependencies?: string[];
+  devDependencies?: string[];
+  skip?: boolean;
+}
+
+export interface TSCodeSample extends BaseCodeSample {
+  ts: string;
+}
+export interface JSCodeSample extends BaseCodeSample {
+  js: string;
+}
+export type CodeSample = TSCodeSample|JSCodeSample;
+
+export function isTSCodeSample(sample: CodeSample): sample is TSCodeSample {
+  return !!(sample as TSCodeSample).ts;
+}
+
+export function isJSCodeSample(sample: CodeSample): sample is JSCodeSample {
+  return !!(sample as JSCodeSample).js;
+}
+
+export interface Dependencies {
+  spawn: typeof spawnP;
+  mkdir: typeof mkdirP;
+  writeFile: typeof writeFileP;
+}
+
+const DEFAULT_DEPENDENCIES = {
+  spawn: spawnP,
+  mkdir: mkdirP,
+  rimraf: rimrafP,
+  tmpDir: tmpDirP,
+  writeFile: writeFileP
+};
+
+interface TestOptions {
+  sample: CodeSample;
+  packageDir?: string;
+  deps?: Partial<Dependencies>;
+}
+
+export async function pack(
+    packageDir: string, targetDir: string): Promise<string> {
+  const packageTarball = path.join(targetDir, 'module-under-test.tgz');
+  const files = await packlist({path: packageDir});
+  await tar.create(
+      {
+        prefix: 'package/',
+        cwd: packageDir,
+        file: packageTarball,
+        gzip: true
+        // tslint:disable-next-line:no-any @types/tar is missing types for
+        // prefix in CreateOptions. TODO: submit fix to @types/tar.
+      } as any,
+      files);
+  return packageTarball;
+}
+
+export async function test(options: TestOptions) {
+  const deps = {...DEFAULT_DEPENDENCIES, ...options.deps};
+  const moduleUnderTest = options.packageDir || process.cwd();
+  let output = '';
+  const installDir = await deps.tmpDir();
+
+  try {
+    const tarball = await pack(moduleUnderTest, installDir);
+    await installInTarget(tarball);
+    await (
+        isTSCodeSample(options.sample) ? checkTS(installDir, options.sample) :
+                                         checkJS(installDir, options.sample));
+  } catch (err) {
+    err.output = output;
+    throw err;
+  } finally {
```

Comment:
Will TypeScript transpile this down into something useful?  Otherwise, this will only work on Node.js 10 and up :)
https://node.green/#ES2018-features-Promise-prototype-finally

Reply:
This is a vanilla `try/catch/finally`, it doesn't have anything to do with `Promises`.